File: 1-two-sum/1-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

/*
data structure: POJO with target - nums[i] as key, i as value
*/

var twoSum = function(nums, target) {
    const mem = {};
    
    for (let i = 0; i < nums.length; i++) {
        if (mem[nums[i]] !== undefined) return [mem[nums[i]], i];
        mem[target - nums[i]] = i;
    }
};

File: 1-two-sum/1-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        data = {}
        for i in range(len(nums)):
            if nums[i] in data:
                return [data[nums[i]], i]
            else:
                data[target - nums[i]] = i